bridgerelayer/chains/substrate/streamer.go

```diff
@@ -1,107 +1,177 @@
-// 				retry--
-// 				time.Sleep(BlockRetryInterval)
-// 				continue
-// 			}
-
-// 			// 1. Get transactions from block...
-
-// 			// 2. Send transaction and block to Router for packaging and relay...
-
-// 			currentBlock++
-// 			retry = BlockRetryLimit
-// 		}
-// 	}
-// }
+import (
+	"fmt"
+	"time"
+	"bytes"
+
+	log "github.com/sirupsen/logrus"
+	types "github.com/snowfork/go-substrate-rpc-client/types"
+	"github.com/snowfork/go-substrate-rpc-client/scale"
+	"github.com/snowfork/polkadot-ethereum/bridgerelayer/chains/ethereum"
+
+)
+
+// Streamer streams Substrate events
+type Streamer struct {
+	Core         *Core
+	EthRouter    *ethereum.Router
+	WebsocketURL string
+	BlockRetryLimit uint
+	BlockRetryInterval time.Duration
+}
+
+// NewStreamer returns a new substrate transaction streamer
+func NewStreamer(core *Core, er *ethereum.Router, websocketURL string, blockRetryLimit uint, blockRetryInterval uint) *Streamer {
+	return &Streamer{
+		Core:         core,
+		EthRouter:    er,
+		WebsocketURL: websocketURL,
+		BlockRetryLimit: blockRetryLimit,
+		BlockRetryInterval: time.Duration(blockRetryInterval) * time.Second,
+	}
+}
+
+// Start the streamer
+func (ss *Streamer) Start() error {
+	// Check whether latest is less than starting block
+	header, err := ss.Core.API.RPC.Chain.GetHeaderLatest()
+	if err != nil {
+		return err
+	}
+	if uint64(header.Number) < ss.Core.StartBlock {
+		return fmt.Errorf("starting block (%d) is greater than latest known block (%d)", ss.Core.StartBlock, header.Number)
+	}
+
+	err = ss.SubscribeBlocks(int(ss.BlockRetryLimit))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SubscribeBlocks ...
+func (ss *Streamer) SubscribeBlocks(blockRetryLimit int) error {
+	var currentBlock = ss.Core.StartBlock
+	var retry = blockRetryLimit
+	for {
+		select {
+		default:
+			// No more retries, go to next block
+			if retry == 0 {
+				log.Error("No more retries")
+				return nil
+			}
+
+			// Get block hash
+			finalizedHash, err := ss.Core.API.RPC.Chain.GetFinalizedHead()
+			if err != nil {
+				log.Error("Failed to fetch head hash", err)
+				retry--
+				time.Sleep(ss.BlockRetryInterval)
+				continue
+			}
+
+			// Get block header
+			finalizedHeader, err := ss.Core.API.RPC.Chain.GetHeader(finalizedHash)
+			if err != nil {
+				log.Error("Failed to fetch finalized header", err)
+				retry--
+				time.Sleep(ss.BlockRetryInterval)
+				continue
+			}
+
+			// Sleep if the block we want comes after the most recently finalized block
+			if currentBlock > uint64(finalizedHeader.Number) {
+				log.WithFields(log.Fields{
+					"target": currentBlock,
+					"latest": finalizedHeader.Number,
+				}).Info("Block not yet finalized")
+				time.Sleep(ss.BlockRetryInterval)
+				continue
+			}
+
+			// Get hash for latest block, sleep and retry if not ready
+			hash, err := ss.Core.API.RPC.Chain.GetBlockHash(currentBlock)
+			if err != nil && err.Error() == "required result to be 32 bytes, but got 0" {
```

Comment:
Can you add a comment when this error is expected to happen?

Reply:
I've cut these issues to track this feedback for implementation in M5:
* https://github.com/Snowfork/polkadot-ethereum/issues/62
* https://github.com/Snowfork/polkadot-ethereum/issues/63